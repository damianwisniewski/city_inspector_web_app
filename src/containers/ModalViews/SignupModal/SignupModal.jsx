import React, { Component } from 'react'

import './SignupModal.scss'

import { Requester } from '../../../services/requester/requester'

import { Form, Input, Group, Checkbox, Radio } from '../../../components/FormComponents'
import Button from '../../../components/CommonComponents/Button/Button'
import Status from '../../../components/CommonComponents/Status/Status'
import RequestStatus from '../../../components/CommonComponents/RequestStatus/RequestStatus'

class SignupModal extends Component {
	constructor(props) {
		super(props)

		this.state = {
			requestState: 'initial',
			email: '',
			password: '',
			repeatedPassword: '',
			emailAgreement: '',
			gender: '',
			nickname: '',
			firstname: '',
			lastname: '',
			city: '',

			dataToSend: {},
		}

		this.repeatPassInput = React.createRef()
	}

	comparePasswords = () => {
		if (this.state.password !== this.state.repeatedPassword) {
			this.repeatPassInput.setCustomValidity('Hasła muszą być identyczne!')
		} else {
			this.repeatPassInput.setCustomValidity('')
		}
	}

	registerUser = e => {
		e.preventDefault()

		this.setState({ requestState: 'pending' })

		// TODO: remove empty values if happends in dataToSend (create helper function, it will be usefull in other places also)
		Requester.send('registerUser', { body: this.state.dataToSend })
			.then(() => {
				this.setState({ requestState: 'succeeded' })
			})
			.catch(() => {
				this.setState({ requestState: 'failed' })
			})
	}

	handleInputChanges = e => {
		const field = e.target.dataset.type
		const isPassInput = field === 'password' || field === 'repeatedPassword'
		const isAgreementField = field === 'emailAgreement'

		const property = {
			[field]: isAgreementField ? (e.target.checked ? 'Y' : 'N') : e.target.value,
		}

		this.setState({
			...property,
			dataToSend: { ...this.state.dataToSend, ...property },
		})

		if (isPassInput) {
			this.comparePasswords()
		}
	}

	render() {
		const { closeModal } = this.props
		const {
			city,
			email,
			firstname,
			lastname,
			nickname,
			password,
			gender,
			emailAgreement,
			repeatedPassword,
			requestState,
		} = this.state

		return (
			<div id='SignupModal' className='signup'>
				<h2 className='signup__header'>Rejestracja</h2>
				{/* <Status id='signup-status' type='error' message='Niepoprawny email lub hasło' /> */}
				<p className='signup__description'>Wpisz następujące dane aby się zarejestrować</p>
				<Form className='signup__form' onSubmit={this.registerUser}>
					{/* Required data */}
					<div className='signup__required'>
						<Input
							id='email-field'
							data-type='email'
							onChange={this.handleInputChanges}
							value={email}
							label='E-mail*'
							placeholder='Podaj swój adres email...'
							labelType='floating'
							type='email'
							pattern='.+@.+'
							autoComplete='email'
							required
						/>
						<Input
							id='password-field'
							data-type='password'
							onChange={this.handleInputChanges}
							onBlur={this.comparePasswords}
							value={password}
							label='Hasło*'
							placeholder='Podaj hasło...'
							labelType='floating'
							type='password'
							name='password'
							autoComplete='new-password'
							minLength='6'
							maxLength='12'
							required
						/>
						<Input
							inputRef={ref => (this.repeatPassInput = ref)}
							id='repeat-password-field'
							data-type='repeatedPassword'
							onChange={this.handleInputChanges}
							onBlur={this.comparePasswords}
							value={repeatedPassword}
							label='Powtórz hasło*'
							placeholder='Powtórz hasło...'
							labelType='floating'
							type='password'
							autoComplete='new-password'
							minLength='6'
							maxLength='12'
							required
						/>
						<Input
							id='nickname-field'
							data-type='nickname'
							onChange={this.handleInputChanges}
							value={nickname}
							label='Nickname*'
							placeholder='Podaj swój adres email...'
							labelType='floating'
							type='text'
							required
						/>
					</div>
					{/* Optionals data */}
					<Group name='opcjonalnie'>
						<Input
							id='name-field'
							data-type='firstname'
							onChange={this.handleInputChanges}
							value={firstname}
							label='Imię'
							placeholder='Podaj swoje imię...'
							labelType='floating'
							type='text'
							autoComplete='given-name'
						/>
						<Input
							id='lastname-field'
							data-type='lastname'
							onChange={this.handleInputChanges}
							value={lastname}
							label='Nazwisko'
							placeholder='Podaj swoje nazwisko...'
							labelType='floating'
							type='text'
							autoComplete='family-name'
						/>
						<Input
							id='city-field'
							data-type='city'
							onChange={this.handleInputChanges}
							value={city}
							label='Miasto'
							placeholder='Podaj swoje miasto...'
							labelType='floating'
							type='text'
							name='city'
							autoComplete='address-level2'
						/>
						<Group name='Płeć' onChange={this.handleInputChanges}>
							<Radio
								data-type='gender'
								name='gender'
								value='M'
								label='Mężczyzna'
								checked={gender === 'M'}
							/>
							<Radio
								data-type='gender'
								name='gender'
								value='F'
								label='Kobieta'
								checked={gender === 'F'}
							/>
						</Group>
					</Group>
					{/* Agreements checkbox */}
					<div className='signup__agreements-wrapper'>
						<Checkbox
							id='emailAgreement'
							data-type='emailAgreement'
							name='emailAgreement'
							checked={emailAgreement === 'Y'}
							onChange={this.handleInputChanges}
							label='Czy wyrażasz zgodę na otrzymywanie informacji mailowych od serwisu "City Inspector", dotyczących zmian w serwisie oraz aktualizacji zgłoszeń.'
						/>
					</div>
					<div className='signup__request-status'>
						<RequestStatus size='small' requestState={requestState}>
							<Status
								id='login-status'
								type='correct'
								message='Konto zostało utworzone, możesz się zalogować'
							/>
						</RequestStatus>
					</div>
					{/* Buttons*/}
					<div className='signup__buttons-wrapper'>
						<Button id='inform-close-button' color='white' onClick={closeModal}>
							Anuluj
						</Button>
						<Button id='submit-button' type='submit' color='blue'>
							Zarejestruj
						</Button>
					</div>
				</Form>
			</div>
		)
	}
}

export default SignupModal
