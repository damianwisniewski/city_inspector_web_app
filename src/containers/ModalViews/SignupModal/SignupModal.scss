@import '../../../assets/styleHelpers/index';

.signup {
	position: relative;

	width: 450px;
	min-height: 320px;

	text-align: center;

	@include phone-size {
		width: unset;
	}

	&__request-status {
		min-width: 100%;
		height: 25px;

		& p[class^='status'] {
			margin: 0;
		}
	}

	&__header {
		margin: 15px 10px;
		padding: 10px;

		color: $light_gray;
		font-size: 20px;
		font-family: $secondary_font;
		text-shadow: $text-shadow;

		background-color: $primary_blue;
		border-radius: 5px;
		box-shadow: $inset-shadow;
	}

	&__description {
		margin: 10px;

		color: $dark_gray;
		font-size: 15px;
	}

	&__form {
		display: flex;
		flex-direction: column;
	}

	&__required {
		margin-bottom: 20px;
	}

	&__buttons-wrapper {
		margin-top: 15px;
		margin-bottom: 10px;
	}

	&__agreements-wrapper {
		min-width: 100%;
		padding: 10px;

		text-align: center;
	}
}
