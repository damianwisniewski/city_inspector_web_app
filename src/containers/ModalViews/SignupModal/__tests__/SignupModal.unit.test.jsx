import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import SignupModal from '../SignupModal'

configure({ adapter: new Adapter() })

describe('<SignupModal />', () => {
	test('should render correctly SignupModal', () => {
		const SignupModalElement = shallow(<SignupModal />)
		expect(toJson(SignupModalElement)).toMatchSnapshot()
	})
})
