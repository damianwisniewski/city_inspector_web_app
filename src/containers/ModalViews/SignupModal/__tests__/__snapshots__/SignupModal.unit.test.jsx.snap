// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupModal /> should render correctly SignupModal 1`] = `
<div
  className="signup"
  id="SignupModal"
>
  <h2
    className="signup__header"
  >
    Rejestracja
  </h2>
  <p
    className="signup__description"
  >
    Wpisz następujące dane aby się zarejestrować
  </p>
  <Form
    className="signup__form"
    onSubmit={[Function]}
  >
    <div
      className="signup__required"
    >
      <Input
        autoComplete="email"
        data-type="email"
        labelType='floating'
        id="email-field"
        inputRef={[Function]}
        label="E-mail*"
        onChange={[Function]}
        pattern=".+@.+"
        placeholder="Podaj swój adres email..."
        required={true}
        type="email"
        value=""
      />
      <Input
        autoComplete="new-password"
        data-type="password"
        labelType='floating'
        id="password-field"
        inputRef={[Function]}
        label="Hasło*"
        maxLength="12"
        minLength="6"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Podaj hasło..."
        required={true}
        type="password"
        value=""
      />
      <Input
        autoComplete="new-password"
        data-type="repeatedPassword"
        labelType='floating'
        id="repeat-password-field"
        inputRef={[Function]}
        label="Powtórz hasło*"
        maxLength="12"
        minLength="6"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Powtórz hasło..."
        required={true}
        type="password"
        value=""
      />
      <Input
        data-type="nickname"
        labelType='floating'
        id="nickname-field"
        inputRef={[Function]}
        label="Nickname*"
        onChange={[Function]}
        placeholder="Podaj swój adres email..."
        required={true}
        type="text"
        value=""
      />
    </div>
    <Group
      direction="column"
      name="opcjonalnie"
    >
      <Input
        autoComplete="given-name"
        data-type="firstname"
        labelType='floating'
        id="name-field"
        inputRef={[Function]}
        label="Imię"
        onChange={[Function]}
        placeholder="Podaj swoje imię..."
        type="text"
        value=""
      />
      <Input
        autoComplete="family-name"
        data-type="lastname"
        labelType='floating'
        id="lastname-field"
        inputRef={[Function]}
        label="Nazwisko"
        onChange={[Function]}
        placeholder="Podaj swoje nazwisko..."
        type="text"
        value=""
      />
      <Input
        autoComplete="address-level2"
        data-type="city"
        labelType='floating'
        id="city-field"
        inputRef={[Function]}
        label="Miasto"
        name="city"
        onChange={[Function]}
        placeholder="Podaj swoje miasto..."
        type="text"
        value=""
      />
    </Group>
    <div
      className="signup__agreements-wrapper"
    >
      <Checkbox
        id="agreement_1"
        label="Czy wyrażasz zgodę na otrzymywanie informacji mailowych od serwisu \\"City Inspector\\", dotyczących zmian w serwisie oraz aktualizacji zgłoszeń."
      />
      <Checkbox
        id="agreement_2"
        label="Czy twoje dane takie jak e-mail, imię i nazwisko oraz miasto, mają być widocznie dla innych zarejestowanych użytkowników."
      />
    </div>
    <RequestStatus
      direction="horizontal"
      errorMessage="Przepraszamy coś poszło nie tak!"
      requestState="initial"
      size="small"
    >
      <Status
        id="login-status"
        message="Konto zostało utworzone, możesz się zalogować"
        position="horizontal"
        type="correct"
      />
    </RequestStatus>
    <div
      className="signup__buttons-wrapper"
    >
      <Button
        category="button"
        className=""
        color="white"
        id="inform-close-button"
        onClick={[Function]}
        type="button"
      >
        Anuluj
      </Button>
      <Button
        category="button"
        className=""
        color="blue"
        id="submit-button"
        onClick={[Function]}
        type="submit"
      >
        Zarejestruj
      </Button>
    </div>
  </Form>
</div>
`;
